# Project - Amazon Recomender System

In [1]:
import numpy as np 
import pandas as pd 
import sklearn

In [2]:
data=pd.read_csv(r"C:\Users\ilaria\Desktop\ilariadata\amazontest.csv")

In [3]:
data.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0  A3R5OBKS7OM2IR     5.0     5.0     NaN     NaN     NaN     NaN     NaN   
1   AH3QC2PC1VTGP     NaN     NaN     2.0     NaN     NaN     NaN     NaN   
2  A3LKP6WPMP9UKX     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
3   AVIY68KEPQ5ZD     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
4  A1CV1WROP5KTTW     NaN     NaN     NaN     NaN     5.0     NaN     NaN   

   Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
1     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
2     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
3     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   

   Movie202  Movie203  Movie204  Movie205  Movie206  
0       NaN       NaN       NaN       NaN       NaN  
1       NaN       NaN       NaN       NaN       NaN  
2       NaN       NaN       NaN       NaN       NaN  
3       NaN       NaN       NaN       NaN       NaN  
4       NaN       NaN       NaN       NaN       NaN  

[5 rows x 207 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4848 entries, 0 to 4847
Columns: 207 entries, user_id to Movie206
dtypes: float64(206), object(1)
memory usage: 7.7+ MB


In [5]:
data.describe()

Movie1  Movie2  Movie3  Movie4     Movie5  Movie6  Movie7  Movie8  \
count     1.0     1.0     1.0     2.0  29.000000     1.0     1.0     1.0   
mean      5.0     5.0     2.0     5.0   4.103448     4.0     5.0     5.0   
std       NaN     NaN     NaN     0.0   1.496301     NaN     NaN     NaN   
min       5.0     5.0     2.0     5.0   1.000000     4.0     5.0     5.0   
25%       5.0     5.0     2.0     5.0   4.000000     4.0     5.0     5.0   
50%       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   
75%       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   
max       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   

       Movie9  Movie10  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
count     1.0      1.0  ...  5.000000       2.0       1.0  8.000000  3.000000   
mean      5.0      5.0  ...  3.800000       5.0       5.0  4.625000  4.333333   
std       NaN      NaN  ...  1.643168       0.0       NaN  0.517549  1.154701   
min       5.0      5.0  ...  1.000000       5.0       5.0  4.000000  3.000000   
25%       5.0      5.0  ...  4.000000       5.0       5.0  4.000000  4.000000   
50%       5.0      5.0  ...  4.000000       5.0       5.0  5.000000  5.000000   
75%       5.0      5.0  ...  5.000000       5.0       5.0  5.000000  5.000000   
max       5.0      5.0  ...  5.000000       5.0       5.0  5.000000  5.000000   

       Movie202  Movie203  Movie204   Movie205   Movie206  
count  6.000000       1.0  8.000000  35.000000  13.000000  
mean   4.333333       3.0  4.375000   4.628571   4.923077  
std    1.632993       NaN  1.407886   0.910259   0.277350  
min    1.000000       3.0  1.000000   1.000000   4.000000  
25%    5.000000       3.0  4.750000   5.000000   5.000000  
50%    5.000000       3.0  5.000000   5.000000   5.000000  
75%    5.000000       3.0  5.000000   5.000000   5.000000  
max    5.000000       3.0  5.000000   5.000000   5.000000  

[8 rows x 206 columns]

In [6]:
#count the NaNs
print(data.isnull().sum())
#there is no movie with 0 views

user_id        0
Movie1      4847
Movie2      4847
Movie3      4847
Movie4      4846
            ... 
Movie202    4842
Movie203    4847
Movie204    4840
Movie205    4813
Movie206    4835
Length: 207, dtype: int64


### EDA

#### which movies have the maximum views/ratings?

In [7]:

moviemaxview =data[data.columns[1:207]]
moviemaxview = moviemaxview.count()

#create a list with the number of views for each movie. User ID coloumn is excluded.

In [8]:
moviemaxview.sort_values(ascending=False).to_frame()


0
Movie127  2313
Movie140   578
Movie16    320
Movie103   272
Movie29    243
...        ...
Movie68      1
Movie69      1
Movie145     1
Movie71      1
Movie1       1

[206 rows x 1 columns]

Movie127 is the most popular, with 2313 views. Movies68,69,145,71,1 are the 5 less viewd with 1 view each

#### What is the average rating for each movie? Define the top 5 movies with the maximum ratings.


In [9]:

movieavrating =data[data.columns[1:207]]
movieavrating = movieavrating.mean(skipna = True)
movieavrating.sort_values(ascending=False)



Movie1      5.0
Movie55     5.0
Movie131    5.0
Movie132    5.0
Movie133    5.0
           ... 
Movie60     1.0
Movie58     1.0
Movie45     1.0
Movie67     1.0
Movie144    1.0
Length: 206, dtype: float64

 The five movies with the maximum rating are Movie1,55,131,132 and 133 with a rating of 5

In [10]:
movieavrating.describe()

count    206.000000
mean       4.448437
std        0.938545
min        1.000000
25%        4.333333
50%        4.961538
75%        5.000000
max        5.000000
dtype: float64

In [11]:
movieavrating['Movie127'].mean()
#the most viewed movie has an average rating of 4.11

4.11197578901859

In [12]:
#swap rows and columns to add average ratings and number of ratings
data_t = data.T
#using the first row as header
data_t.columns = data_t.iloc[0] 
data_t = data_t.drop(data_t.index[0])
# add a column with average rating and number of views for each movie
data_t['AverageRating'] = movieavrating
data_t['ViewsNumber'] = moviemaxview
data_t.head()

user_id A3R5OBKS7OM2IR AH3QC2PC1VTGP A3LKP6WPMP9UKX AVIY68KEPQ5ZD  \
Movie1               5           NaN            NaN           NaN   
Movie2               5           NaN            NaN           NaN   
Movie3             NaN             2            NaN           NaN   
Movie4             NaN           NaN              5             5   
Movie5             NaN           NaN            NaN           NaN   

user_id A1CV1WROP5KTTW AP57WZ2X4G0AA A3NMBJ2LCRCATT A5Y15SAOMX6XA  \
Movie1             NaN           NaN            NaN           NaN   
Movie2             NaN           NaN            NaN           NaN   
Movie3             NaN           NaN            NaN           NaN   
Movie4             NaN           NaN            NaN           NaN   
Movie5               5             2              5             2   

user_id A3P671HJ32TCSF A3VCKTRD24BG7K  ... A15W9FRLBUXJNS AQ4JA4RFC5QHI  \
Movie1             NaN            NaN  ...            NaN           NaN   
Movie2             NaN            NaN  ...            NaN           NaN   
Movie3             NaN            NaN  ...            NaN           NaN   
Movie4             NaN            NaN  ...            NaN           NaN   
Movie5               5              5  ...            NaN           NaN   

user_id A1HY5DOJCVB8MG A1IMQ9WMFYKWH5 A1KLIKPUF5E88I A5HG6WFZLO10D  \
Movie1             NaN            NaN            NaN           NaN   
Movie2             NaN            NaN            NaN           NaN   
Movie3             NaN            NaN            NaN           NaN   
Movie4             NaN            NaN            NaN           NaN   
Movie5             NaN            NaN            NaN           NaN   

user_id A3UU690TWXCG1X AI4J762YI6S06 AverageRating ViewsNumber  
Movie1             NaN           NaN      5.000000           1  
Movie2             NaN           NaN      5.000000           1  
Movie3             NaN           NaN      2.000000           1  
Movie4             NaN           NaN      5.000000           2  
Movie5             NaN           NaN      4.103448          29  

[5 rows x 4850 columns]

In [13]:
data_t.describe() 

user_id  AverageRating  ViewsNumber
count       206.000000   206.000000
mean          4.448437    24.271845
std           0.938545   168.937841
min           1.000000     1.000000
25%           4.333333     1.000000
50%           4.961538     2.000000
75%           5.000000     5.000000
max           5.000000  2313.000000

### Recommendation Model: Some of the movies hadn’t been watched and therefore, are not rated by the users.  Netflix would like to take this as an opportunity and build a machine learning recommendation algorithm which provides the ratings for each of the users.


* Divide the data into training and test data
* Build a recommendation model on training data
* Make predictions on the test data

In [14]:
from surprise import Reader
from surprise import accuracy
from surprise import Dataset
from surprise.model_selection import train_test_split
from surprise import SVD
from surprise.model_selection import cross_validate
from surprise.model_selection import GridSearchCV

In [15]:
data_melt = data.melt(id_vars=data.columns[0], value_vars=data.columns[1:], var_name = 'movie',value_name = 'ratings')
data_melt

user_id     movie  ratings
0       A3R5OBKS7OM2IR    Movie1      5.0
1        AH3QC2PC1VTGP    Movie1      NaN
2       A3LKP6WPMP9UKX    Movie1      NaN
3        AVIY68KEPQ5ZD    Movie1      NaN
4       A1CV1WROP5KTTW    Movie1      NaN
...                ...       ...      ...
998683  A1IMQ9WMFYKWH5  Movie206      5.0
998684  A1KLIKPUF5E88I  Movie206      5.0
998685   A5HG6WFZLO10D  Movie206      5.0
998686  A3UU690TWXCG1X  Movie206      5.0
998687   AI4J762YI6S06  Movie206      5.0

[998688 rows x 3 columns]

In [16]:
data_melt['ratings'].mean()

4.3856

In [17]:
reader = Reader(rating_scale=(1,5))
data_r = Dataset.load_from_df(data_melt.fillna(data_melt['ratings'].mean()), reader=reader)

In [18]:
data_r

In [19]:
#divide train and test
train,test= train_test_split(data_r, test_size=0.25)

In [20]:
alg=SVD()

In [21]:
#applying the SVD to the train set
alg.fit(train)

In [22]:
pred = alg.test(test)

In [23]:
accuracy.rmse(pred)

RMSE: 0.0861


0.0860731138076564

In [24]:
#it seems a faily good approximation

us_id = 'A3R5OBKS7OM2IR'
movie_id = 'Movie1'
r_true = 5.0
alg.predict(us_id,movie_id,r_true, verbose=True)

user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 4.39   {'was_impossible': False}


Prediction(uid='A3R5OBKS7OM2IR', iid='Movie1', r_ui=5.0, est=4.390714783226707, details={'was_impossible': False})

In [25]:
#create a function

def repeat(alg_type, frame, min, max):
    reader = Reader(rating_scale=(1,5))
    
    data_r = Dataset.load_from_df(frame, reader=reader)

    alg=alg_type
    
    print(cross_validate(alg, data_r, measures = ['RMSE','MAE'],cv=5, verbose = True))
    
    print("#"*10)
    
    us_id = 'A3R5OBKS7OM2IR'
    movie_id = 'Movie1'
    r_true = 5.0
    print(alg.predict(us_id,movie_id,r_true, verbose=True))
    
    print("#"*20)
    
    print()

In [26]:
#3 different models to compare

repeat(SVD(),data_melt.fillna(data_melt['ratings'].mean()),1,5)
repeat(SVD(),data_melt.fillna(data_melt['ratings'].median()),1,5)
repeat(SVD(),data_melt.fillna(0),1,5)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.0806  0.0842  0.0877  0.0867  0.0890  0.0856  0.0030  
MAE (testset)     0.0093  0.0096  0.0097  0.0097  0.0103  0.0097  0.0003  
Fit time          113.63  103.21  106.88  103.80  103.86  106.27  3.89    
Test time         5.28    4.82    5.49    5.48    5.52    5.32    0.26    
{'test_rmse': array([0.08063069, 0.084188  , 0.08768947, 0.08667817, 0.08898598]), 'test_mae': array([0.00926498, 0.00964063, 0.00966895, 0.0097057 , 0.01034698]), 'fit_time': (113.62621808052063, 103.20858311653137, 106.88007593154907, 103.79779362678528, 103.86118483543396), 'test_time': (5.281883001327515, 4.822408676147461, 5.493209600448608, 5.476609468460083, 5.524409770965576)}
##########
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 4.39   {'was_impossible': False}
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 4.39   {'was_impossible

In [27]:
param_grid = {'n_epochs':[20,30],
              'lr_all':[0.005,0.01],
              'n_factors':[50,100] }

In [28]:
gs = GridSearchCV(SVD, param_grid, measures=['rmse','mae'],cv=5)
gs.fit(data_r)

In [29]:
gs.best_score

{'rmse': 0.08473152041639709, 'mae': 0.008434165369834847}